content/en/articles/tdd/index.md

```diff
@@ -0,0 +1,294 @@
+
+In theory, this method requires the attendance of at least two different coders,
+one to write the tests and the other write the tested code. That allowing issues
+linked to subjectivity.
+
+In practice things are more complicated, sometimes we are developing alone or
+we are writing by ourselves the tests that ensure the integrity of a new feature
+in a collaborative project.
+
+> Whatever happens, a low-efficient test is more valuable that no test at all.
+> The aim being to be used to write some and to be objective in their writing.
+
+TDD is becoming increasingly popular and require the willing of everyone to
+become mainstream. Any developer concerned about his environment and his legacy
+have to ask himself seriously.  
+Test frameworks, guides and documentations about it are flourishing so you can
+get on board easily.
+
+We can split TDD in 5 distinct steps :
+
+1. Write a test.
+2. Make it fail.
+3. Write the **suitable** code to make the test pass.
+4. Check that the test pass.
+5. Optimize the code and check that there is no regression.
+
+
+To simplify this logic we can group these 5 steps in 3 great principles:
+
+- **Test first**, which corresponds to the first two steps.
+- **Make it work**, that covers point 3 and 4.
+- **Make it better**, which is none other than the 5th step.
+
+Bill Wake define these three points in a methodology called 
+[3A](http://xp123.com/articles/3a-arrange-act-assert/) 
+standing for Arrange, Act, Assert.  
+He insist on the fact that the order isn't immutable, as such,
+_assertion_ can come before _action_, etc...
+
+- **Arranger** : Set up phase of the test environment in which we declare 
+variables and functions. 
+- **Act** : Phase in which we act on our environment by applying mutations. 
+- **Assert** : We make claims about the whole environment 
+(variables, functions, parameters).
+
+## TDD 101
+
+For the following test set we'll use EcmaScript 6 and 
+[`.assert()`](https://developer.mozilla.org/en-US/docs/Web/API/console/assert) 
+method from the browser console: Like so, you can reproduce them at will.
+
+Goal : Write a function `countWords()` that count words.. Duh..
+
+**ITERATION #1** : write and fail of the initial test
+
+First we write a basic assertion.
+```js
+console.assert(countWords('') === 0, 'test 0: text is empty');
+```
+> `Uncaught ReferenceError: countWords is not defined`
+
+After running this snippet the console reject our test. 
+We have to define `countWords()` first.
+
+```js
+const countWords = () => {};
+console.assert(countWords('') === 0, 'test 0: text is empty');
+```
+> `Assertion failed: test 0: text is empty`
+
+Function is now defined and the test fails but the error has changed.  
+We now have to define the core of our function.
+
+```js
+const countWords = text => text || 0;
+console.assert(countWords('') === 0, 'test 0: text is empty');
+```
+> `undefined`
+
+The console isn't outputting nothing, the test has passed.  
```

Comment:
I'm not sure colloquialism has a place in technical writing, and this usage comes across to me as confusing.

> The tests pass because the console hasn't logged an error!